libraries/backend/azure_connection.rb

```diff
@@ -0,0 +1,182 @@
+# frozen_string_literal: true
+require 'rubygems'
+require 'bundler/setup'
+require 'faraday'
+require 'faraday_middleware'
+require 'uri'
+require 'pp'
+require_relative 'helpers'
+
+# Client class to manage the Azure REST API connection.
+#
+# An instance of this class will:
+#   - create a HTTP client.
+#   - read environment variables and gather necessary information for authentication.
+#   - authenticate with Azure Rest API and gets an access token.
+#   - make the access token available to use.
+class AzureConnection
+  @@token_data = Hash.recursive
+  # This will be included in headers for statistical purposes.
+  INSPEC_USER_AGENT = 'pid-18d63047-6cdf-4f34-beed-62f01fc73fc2'
+
+  # @return [String] the resource management endpoint url
+  attr_reader :resource_manager_endpoint_url
+
+  # @return [String] the resource management endpoint api version, e.g. 2020-01-01
+  attr_reader :resource_manager_endpoint_api_version
+
+  # @return [String] the graph api endpoint url
+  attr_reader :graph_api_endpoint_url
+
+  # @return [String] the graph api endpoint api version, e.g. v1.0
+  attr_reader :graph_api_endpoint_api_version
+
+  # @return [Hash] tenant_id, client_id, client_secret, subscription_id
+  attr_reader :credentials
+
+  # Creates a HTTP client.
+  def initialize(client_args)
+    # Validate parameter's type.
+    raise ArgumentError, 'Parameters must be provided in an Hash object.' unless client_args.is_a?(Hash)
+
+    # The valid client args:
+    #   - endpoint: [String]
+    #       azure_cloud, azure_china_cloud, azure_us_government_L4, azure_us_government_L5, azure_german_cloud
+    #   - azure_retry_limit: [Integer] Maximum number of retries (default - 2)
+    #   - azure_retry_backoff: [Integer] Pause in seconds between retries (default - 0)
+    #   - azure_retry_backoff_factor: [Integer] The amount to multiply each successive retry's interval amount by
+    #       in order to provide back-off (default - 1)
+    @client_args = client_args
+
+    raise StandardError, 'Endpoint has to be provided to establish a connection with Azure REST API.' \
+      unless @client_args.key?(:endpoint)
+    @resource_manager_endpoint_url = @client_args[:endpoint].resource_manager_endpoint_url
+    @resource_manager_endpoint_api_version = @client_args[:endpoint].resource_manager_endpoint_api_version
+    @graph_api_endpoint_url = @client_args[:endpoint].graph_api_endpoint_url
+    @graph_api_endpoint_api_version = @client_args[:endpoint].graph_api_endpoint_api_version
+
+    @credentials = {
+      tenant_id: ENV['AZURE_TENANT_ID'],
+      client_id: ENV['AZURE_CLIENT_ID'],
+      client_secret: ENV['AZURE_CLIENT_SECRET'],
+      subscription_id: ENV['AZURE_SUBSCRIPTION_ID'],
+    }
+    # Validate the presence of credentials.
+    unless @credentials.values.compact.size == 4
+      raise UnsuccessfulAPIQuery::MissingCredentials, "All of the following have to be provided: #{@credentials.keys}.\n"\
```

Comment:
Potential re-word:
`"The following credentials must be set in the Environment: #{@credentials.keys}.\n"`